src/randgraphs_sbm.jl

```diff
@@ -0,0 +1,166 @@
+the internal probabilities and sizes differ by blocks.
+"""
+function sbmaffinity(internalp::Vector{Float64}, externalp::Float64, sizes::Vector{Int})
+    numblocks = length(sizes)
+    numblocks == length(internalp) || error("Inconsistent input dimensions: internalp, sizes")
+    B = diagm(internalp) + externalp*(ones(numblocks, numblocks)-I)
+    return B
+end
+
+function StochasticBlockModel(internalp::Float64,
+                              externalp::Float64,
+                              size::Int,
+                              numblocks::Int)
+    sizes = [size for i in 1:numblocks]
+    B = sbmaffinity([internalp for i in 1:numblocks], externalp, sizes)
+    StochasticBlockModel(sizes, B)
+end
+
+function StochasticBlockModel(internalp::Vector{Float64}, externalp::Float64, sizes::Vector{Int})
+    B = sbmaffinity(internalp, externalp, sizes)
+    return StochasticBlockModel(sizes, B)
+end
+
+
+biclique = ones(2,2) - eye(2)
+
+"""construct the affinity matrix for a near bipartite SBM.
+between is the affinity between the two parts of each bipartite community
+intra is the probability of an edge within the parts of the partitions.
+
+This is a specific type of SBM with k/2 blocks each with two halves.
+Each half is connected as a random bipartite graph with probability `intra`
+The blocks are connected with probability `between`.
+"""
+function nearbipartiteaffinity(sizes::Vector{Int}, between::Float64, intra::Float64)
+    numblocks = round(Int, length(sizes)/2)
+    return kron(between*eye(numblocks), biclique) + eye(2numblocks)*intra
+end
+
+"""Return a generator for edges from a stochastic block model near bipartite graph."""
+function nearbipartiteaffinity(sizes::Vector{Int}, between::Float64, inter::Float64, noise)
+    B = nearbipartiteaffinity(sizes, between, inter) + noise
+    # info("Affinities are:\n$B")#, file=stderr)
+    return B
+end
+
+function nearbipartiteSBM(sizes, between, inter, noise)
+    return StochasticBlockModel(sizes, nearbipartiteaffinity(sizes, between, inter, noise))
+end
+
+
+"""random_pair: generates a stream of random pairs in 1:n"""
+function random_pair(n::Int)
+    while true
+        produce( rand(1:n), rand(1:n) )
+    end
+end
+
+
+"""Take an infinite sample from the sbm.
+Pass to graph(edgestream, numedges) to get a Graph object.
+"""
+function sample(sbm::StochasticBlockModel)
+    pairs = @task random_pair(sbm.n)
+    for (i,j) in pairs
+    	if i == j
+            continue
+        end
+        p = sbm.affinities[sbm.nodemap[i], sbm.nodemap[j]]
+        if rand() < p
+            produce(i, j)
+        end
+    end
+end
+
+"""convert a stream of edges produced by sample into a graph"""
+function graph(edgestream, sizes, numedges::Int)
+    g = Graph(sum(sizes))
+    # println(g)
+    count = 1
+    for (i,j) in edgestream
+        # print("$count, $i,$j\n")
+        if !has_edge(g,i,j)
+            add_edge!(g,Edge(i,j))
```

Comment:
You might be able to do this more efficiently when we implement #246 :)

Also, you can just do `add_edge!(g, i, j)`.